File: src/Application.php
<?php

declare(strict_types=1);

namespace ServerLens;

use ServerLens\Audit\AuditLogger;
use ServerLens\Auth\RateLimiter;
use ServerLens\Auth\TokenAuth;
use ServerLens\Mcp\Server;
use ServerLens\Module\ConfigReader;
use ServerLens\Module\DbQuery;
use ServerLens\Module\LogReader;
use ServerLens\Module\SystemInfo;
use ServerLens\Transport\SseTransport;
use ServerLens\Transport\StdioTransport;
use ServerLens\Transport\TransportInterface;

final class Application
{
    private Config $config;
    private Server $mcpServer;
    private TransportInterface $transport;

    public function __construct(string $configPath)
    {
        $this->config = Config::load($configPath);
        $this->boot();
    }

    public function run(): void
    {
        fwrite(STDERR, "[ServerLens] Starting server...\n");
        fwrite(STDERR, "[ServerLens] Transport: {$this->config->getTransport()}\n");

        $this->transport->onMessage(function (array $message, string $clientIp): ?array {
            return $this->mcpServer->handleMessage($message, $clientIp);
        });

        $this->transport->start();
    }

    private function boot(): void
    {
        $audit = null;
        if ($this->config->isAuditEnabled()) {
            $audit = new AuditLogger(
                $this->config->getAuditPath(),
                $this->config->shouldLogParams(),
            );
        }

        $rateLimiter = new RateLimiter(
            $this->config->getRequestsPerMinute(),
            $this->config->getMaxConcurrent(),
        );

        $this->mcpServer = new Server($audit, $rateLimiter);

        $this->registerModules();
        $this->createTransport();
    }

    private function registerModules(): void
    {
        if (!empty($this->config->getLogSources())) {
            $this->mcpServer->registerModule(new LogReader($this->config));
            fwrite(STDERR, "[ServerLens] Module loaded: LogReader\n");
        }

        if (!empty($this->config->getConfigSources())) {
            $this->mcpServer->registerModule(new ConfigReader($this->config));
            fwrite(STDERR, "[ServerLens] Module loaded: ConfigReader\n");
        }

        if (!empty($this->config->getDatabaseConnections())) {
            $this->mcpServer->registerModule(new DbQuery($this->config));
            fwrite(STDERR, "[ServerLens] Module loaded: DbQuery\n");
        }

        if ($this->config->isSystemEnabled()) {
            $this->mcpServer->registerModule(new SystemInfo($this->config));
            fwrite(STDERR, "[ServerLens] Module loaded: SystemInfo\n");
        }
    }

    private function createTransport(): void
    {
        $transportType = $this->config->getTransport();

        if ($transportType === 'sse') {
            $auth = new TokenAuth($this->config);
            $this->transport = new SseTransport(
                $this->config->getServerHost(),
                $this->config->getServerPort(),
                $auth,
            );
        } else {
            $this->transport = new StdioTransport();
        }
    }
}

File: serverlens/application.py
from __future__ import annotations

import signal
import sys

from serverlens.audit.audit_logger import AuditLogger
from serverlens.auth.rate_limiter import RateLimiter
from serverlens.auth.token_auth import TokenAuth
from serverlens.config import Config
from serverlens.mcp.server import Server
from serverlens.module.config_reader import ConfigReader
from serverlens.module.config_suggest import ConfigSuggest
from serverlens.module.db_query import DbQuery
from serverlens.module.journal_reader import JournalReader
from serverlens.module.log_reader import LogReader
from serverlens.module.system_info import SystemInfo
from serverlens.transport.base import TransportInterface
from serverlens.transport.sse import SseTransport
from serverlens.transport.stdio import StdioTransport


class Application:
    def __init__(self, config_path: str) -> None:
        self._config_path = config_path
        self._config = Config.load(config_path)
        self._mcp_server: Server
        self._transport: TransportInterface
        self._boot()

    def run(self) -> None:
        print(f"[ServerLens] Starting server...", file=sys.stderr)
        print(f"[ServerLens] Transport: {self._config.get_transport()}", file=sys.stderr)

        self._install_reload_handler()

        self._transport.on_message(
            lambda message, client_ip: self._mcp_server.handle_message(message, client_ip)
        )
        self._transport.start()

    def _install_reload_handler(self) -> None:
        # SIGHUP → re-read config and re-register modules without dropping the
        # process. Matters for the long-lived SSE service; harmless elsewhere.
        # (stdio sessions are short-lived and always read fresh config anyway.)
        if not hasattr(signal, "SIGHUP"):
            return
        try:
            signal.signal(signal.SIGHUP, lambda *_: self.reload())
        except (ValueError, OSError):
            pass  # not on the main thread — skip

    def reload(self) -> None:
        try:
            new_config = Config.load(self._config_path)
        except Exception as e:
            print(f"[ServerLens] Reload skipped — invalid config: {e}", file=sys.stderr)
            return
        self._config = new_config
        self._mcp_server.reset_modules()
        self._register_modules()
        print("[ServerLens] Configuration reloaded", file=sys.stderr)

    def _boot(self) -> None:
        audit = None
        if self._config.is_audit_enabled():
            audit = AuditLogger(
                self._config.get_audit_path(),
                self._config.should_log_params(),
            )

        rate_limiter = RateLimiter(
            self._config.get_requests_per_minute(),
            self._config.get_max_concurrent(),
        )

        self._mcp_server = Server(audit, rate_limiter)
        self._register_modules()
        self._create_transport()

    def _register_modules(self) -> None:
        if self._config.get_log_sources():
            self._mcp_server.register_module(LogReader(self._config))
            print("[ServerLens] Module loaded: LogReader", file=sys.stderr)

        if self._config.get_config_sources():
            self._mcp_server.register_module(ConfigReader(self._config))
            print("[ServerLens] Module loaded: ConfigReader", file=sys.stderr)

        if self._config.get_database_connections():
            self._mcp_server.register_module(DbQuery(self._config))
            print("[ServerLens] Module loaded: DbQuery", file=sys.stderr)
            self._mcp_server.register_module(ConfigSuggest(self._config))
            print("[ServerLens] Module loaded: ConfigSuggest", file=sys.stderr)

        if self._config.is_system_enabled():
            self._mcp_server.register_module(SystemInfo(self._config))
            print("[ServerLens] Module loaded: SystemInfo", file=sys.stderr)

        if self._config.is_journal_enabled() and self._config.get_allowed_journal_units():
            self._mcp_server.register_module(JournalReader(self._config))
            print("[ServerLens] Module loaded: JournalReader", file=sys.stderr)

    def _create_transport(self) -> None:
        if self._config.get_transport() == "sse":
            auth = TokenAuth(self._config)
            self._transport = SseTransport(
                self._config.get_server_host(),
                self._config.get_server_port(),
                auth,
            )
        else:
            self._transport = StdioTransport()
